### DISCUSSION TITLE
Best way to remove white lines on a pie chart?

### DISCUSSION THREAD

**Discussion by Readof** [ImageMagick/ImageMagick#3573] [Help] [answered]
I'm using a chart-generating external API to generate pie charts that look something like this:

![image](https://user-images.githubusercontent.com/82979076/115627182-69096080-a2b3-11eb-87de-1ab9015b185d.png)

While the API works fine, the final image has some white lines between each slice that I'm trying to remove. Since the API doesn't have any option to remove them, I'm trying to use ImageMagick instead.

For charts with only two slices, something like:
`convert image -fuzz 20% -fill '#b0ffd7' -opaque white file`
works fine:
![image](https://user-images.githubusercontent.com/82979076/115634349-4c265a80-a2be-11eb-8ac3-b146ccfa3ad4.png)

But what I need help with is charts with more than two slices:
![image](https://user-images.githubusercontent.com/82979076/115637814-ed64df00-a2c5-11eb-9db1-14d7abce0fb1.png)
My original method doesn't work, and I'm not sure what I can do. Any help is appreciated.

IM Version: 6.9.10-23 Q16 arm 20190101
On: Raspberry Pi 4

**Comment by fmw42** [comment]
Please show your API code? Are you sure it does not actually draw those white lines? Please show your command line command for multiple section also.

This is not a complete command!

`convert image -fuzz 20% -fill '#b0ffd7' -opaque white`

**Reply by Readof** [reply]
Thanks for the quick response.

I'm using [Google Image Charts](https://developers.google.com/chart/image), and it does indeed draw those white lines. I've looked extensively through their documentation, and as far as I know, there's no way to get rid of them. (I'm aware it's very, very outdated, but it still accepts requests, so I'll continue to use it for as long as I can.)

Here are the URLs I've used:

Two-slice: https://chart.googleapis.com/chart?cht=p&chs=480x360&chd=t:10,5&chf=bg,s,50422700&chco=121a16,b0ffd7&chdls=58806c,20

Five-slice: https://chart.googleapis.com/chart?cht=p&chs=480x360&chd=t:10,5,12,20,1&chf=bg,s,50422700&chco=121a16,b0ffd7&chdls=58806c,20

As for the five-slice pie chart, all I've tried was my original method, which resulted in this:
![image](https://user-images.githubusercontent.com/82979076/115638764-d626f100-a2c7-11eb-92ce-ca6a6a4ea280.png)

-----

The complete command I used was:
`convert 'https://chart.googleapis.com/chart?cht=p&chs=480x360&chd=t:10,5&chf=bg,s,50422700&chco=121a16,b0ffd7&chdls=58806c,20' -fuzz 20% -fill '#b0ffd7' -opaque white PieChart.png`

**Comment by fmw42** [comment]
I do not understand. I thought you were using an Imagemagick API. I do not know what Google Image charts does or uses.  

You appear to be try to change white to some other color using -fuzz. Perhaps the white color is not really white. Have you increased your fuzz value?

Have you considered just using Imagemagick to draw pie charts? See the pie chart example right after https://legacy.imagemagick.org/Usage/draw/#arcs

**Comment by fmw42** [comment, marked as answer]
This mostly works for me. There are a few "grayish" pixels along the edges between colors, but mostly the white is gone.

![chart](https://user-images.githubusercontent.com/13458218/115639830-6a925300-a2ca-11eb-9060-2a67f7a5c934.png)


`convert chart.png -fuzz 45% -fill '#b0ffd7' -opaque white x.png`

![x](https://user-images.githubusercontent.com/13458218/115639863-7ed65000-a2ca-11eb-9ca5-119e45a40dcd.png)

I do not know how to do this for multiple sections, since once you change white to some color, then it might not match the other sections if they are not that color. The only way I can think to do that is to draw a colored line over the white line at each transition. But you would have to know how the diagram was drawn. So you might as well draw it as you would like in Imagemagick or some other tool.

**Reply by Readof** [reply]
Thanks, this is the best method I've tried so far. I'm also going to try to create the charts with ImageMagick.

**Comment by snibgo** [comment]
Connected-components can do the job. See https://imagemagick.org/script/connected-components.php . Windows BAT syntax:
```
magick ^
  piecht.png ^
  -define connected-components:area-threshold=400 ^
  -define connected-components:mean-color=true ^
  -connected-components 4 -channel A -threshold 90% +channel ^
  out.png
```
![x](https://user-images.githubusercontent.com/33812368/115704558-69f9cb00-a363-11eb-9973-13eb4b0c45ce.png)

This removes aliasing, so sloping lines become jagged. Fixing that is harder.
